extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+# workspace/reference extension to LSP
+
+The `workspace/reference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references that code inside of the workspace root makes to code symbols that are external to the workspace root (e.g. dependencies).
+
+Use case: clients of a language server can invoke `workspace/reference` in order to find external references to e.g. imported dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.URI` which is located within the workspace.
+- Excluding any `ReferenceInformation.Name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.Location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace at which the `symbol` has been referenced.
+	 */
+	location: Location;
+
+	/**
+	 * The name of the symbol that is being referenced. For example:
+	 *
+	 * 	(Go) "ServeHTTP"
+	 *  (JS) "render"
+	 *
+	 */
+	name?: string;
+
+	/**
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
```

Comment:
> Why does it need to uniquely identify it? This adds complexity and doesn't seem necessary.

_Some_ level of uniqueness is needed for handling "Top Definitions" on repo landing pages. Otherwise, we would have to ask the user "Can you disambiguate the top 20 defs?" (obviously we wouldn't do this).

> (And it'd require name munging and additional buffering in memory to ensure uniqueness.)

I don't see why you say this? We don't do it for e.g. Go